src/Dialect/ONNX/ONNXOps/ShapeHelper.cpp

```diff
@@ -327,13 +330,12 @@ LogicalResult ONNXBroadcastOpShapeHelper::customComputeShape(
   return success();
 }
 
+// Attempt to rule out broadcasting at compile time, using dim analysis when
+// available (i.e. nonnull). Must be called after computeShape.
 bool ONNXBroadcastOpShapeHelper::hasNoBroadcast(DimAnalysis *dimAnalysis) {
```

Comment:
@AlexandreEichenberger could you add the lit test from #2158 into this patch to make sure it works?

By the way, `hasNoBroadcasting` is still confusing, I don't see how it solves the semantics issue @chentong319 mentioned in #2158. For example, `hasNoBroadcasting` still returns `true` for the case `%0 = "onnx.Add"(%arg0, %arg1) : (tensor<1x1x5xf32>, tensor<5xf32>) -> tensor<1x1x5xf32>`, doesn't it?

Reply:
There is a lit test that check for that specific pattern, and make sure that it get simdized as it detects no broadcast.